fastcrypto-zkp/src/circom.rs

```diff
@@ -1,22 +1,142 @@
+    where
+        S: serde::Serializer,
+    {
+        self.to_string().serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for Bn254FqElement {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let s = std::borrow::Cow::<'de, str>::deserialize(deserializer)?;
+        std::str::FromStr::from_str(&s).map_err(serde::de::Error::custom)
+    }
+}
+
+/// A struct that stores a Bn254 Fr field element as 32 bytes.
+#[derive(Debug, Clone, JsonSchema, Eq, PartialEq)]
+pub struct Bn254FrElement([u8; 32]);
+
+impl Bn254FrElement {
+    /// Returns the unpadded version of the field element. This returns with leading zeros removed.
+    pub fn unpadded(&self) -> &[u8] {
+        let mut buf = self.0.as_slice();
+
+        while !buf.is_empty() && buf[0] == 0 {
+            buf = &buf[1..];
+        }
+
+        // If the value is '0' then just return a slice of length 1 of the final byte
+        if buf.is_empty() {
+            &self.0[31..]
+        } else {
+            buf
+        }
+    }
+
+    /// Returns the padded version of the field element. This returns with leading zeros preserved to 32 bytes.
+    pub fn padded(&self) -> &[u8] {
+        &self.0
+    }
+}
+impl std::str::FromStr for Bn254FrElement {
+    type Err = FastCryptoError;
 
-/// Parse a string as a field element in BN254. Return an `FastCryptoError::InvalidInput` error if
-/// the parsing fails.
-fn parse_field_element(s: &str) -> Result<Fq, FastCryptoError> {
-    s.parse::<Fq>().map_err(|_| FastCryptoError::InvalidInput)
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let big_int = Fr::from_str(s).map_err(|_| FastCryptoError::InvalidInput)?;
+        let be_bytes = big_int.into_bigint().to_bytes_be();
+        be_bytes
+            .try_into()
+            .map_err(|_| FastCryptoError::InvalidInput)
+            .map(Bn254FrElement)
+    }
+}
+
+impl std::fmt::Display for Bn254FrElement {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let big_int = BigUint::from_bytes_be(&self.0);
+        let radix10 = big_int.to_string();
+        f.write_str(&radix10)
+    }
+}
+
+// Bn254FrElement's serialized format is as a radix10 encoded string
+impl Serialize for Bn254FrElement {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        self.to_string().serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for Bn254FrElement {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let s = std::borrow::Cow::<'de, str>::deserialize(deserializer)?;
+        std::str::FromStr::from_str(&s).map_err(serde::de::Error::custom)
+    }
+}
+
+/// Convert Bn254FqElement type to arkworks' Fq.
+fn parse_fq_field_element(s: &Bn254FqElement) -> Result<Fq, FastCryptoError> {
```

Comment:
The `Result` is not needed here or below. It could also be implemented as `Into<Fq>`, but this works too.